e2e/tests/parse_params_test.go

```diff
@@ -207,94 +215,131 @@ func (s *suite) newMsgUpdateParam(
 
 	authority := authtypes.NewModuleAddress(s.granterName).String()
 
-	// TODO_IMPROVE: can this be simplified?
 	switch moduleName {
 	case tokenomicstypes.ModuleName:
-		switch param.typeStr {
-		case "string":
-			msg = proto.Message(&tokenomicstypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &tokenomicstypes.MsgUpdateParam_AsString{
-					AsString: param.value.(string),
-				},
-			})
-		case "int64":
-			msg = proto.Message(&tokenomicstypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &tokenomicstypes.MsgUpdateParam_AsInt64{
-					AsInt64: param.value.(int64),
-				},
-			})
-		case "bytes":
-			msg = proto.Message(&tokenomicstypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &tokenomicstypes.MsgUpdateParam_AsBytes{
-					AsBytes: param.value.([]byte),
-				},
-			})
-		}
+		msg = s.newTokenomicsMsgUpdateParam(authority, param)
 	case prooftypes.ModuleName:
-		switch param.typeStr {
-		case "string":
-			msg = proto.Message(&prooftypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &prooftypes.MsgUpdateParam_AsString{
-					AsString: param.value.(string),
-				},
-			})
-		case "int64":
-			msg = proto.Message(&prooftypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &prooftypes.MsgUpdateParam_AsInt64{
-					AsInt64: param.value.(int64),
-				},
-			})
-		case "bytes":
-			msg = proto.Message(&prooftypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &prooftypes.MsgUpdateParam_AsBytes{
-					AsBytes: param.value.([]byte),
-				},
-			})
-		}
+		msg = s.newProofMsgUpdateParam(authority, param)
 	case sharedtypes.ModuleName:
-		switch param.typeStr {
-		case "string":
-			msg = proto.Message(&sharedtypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &sharedtypes.MsgUpdateParam_AsString{
-					AsString: param.value.(string),
-				},
-			})
-		case "int64":
-			msg = proto.Message(&sharedtypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &sharedtypes.MsgUpdateParam_AsInt64{
-					AsInt64: param.value.(int64),
-				},
-			})
-		case "bytes":
-			msg = proto.Message(&sharedtypes.MsgUpdateParam{
-				Authority: authority,
-				Name:      param.name,
-				AsType: &sharedtypes.MsgUpdateParam_AsBytes{
-					AsBytes: param.value.([]byte),
-				},
-			})
-		}
+		msg = s.newSharedMsgUpdateParam(authority, param)
 	default:
 		err := fmt.Errorf("ERROR: unexpected module name %q", moduleName)
 		s.Fatal(err)
 		panic(err)
 	}
 
+	if msg == nil {
+		err := fmt.Errorf("ERROR: unexpected param type %q for %q module", param.typeStr, moduleName)
+		s.Fatal(err)
+		panic(err)
+	}
+
+	return msg
+}
+
+func (s *suite) newTokenomicsMsgUpdateParam(authority string, param paramAny) (msg proto.Message) {
```

Comment:
In tests, I often appreciate not following DRY, but do you think generics or reflections are worth a shot given where this is headed?

Reply:
I tried to come up with a way to express this using generics but was unsuccessful, the fact that we need to assign to variable fields on the underlying struct makes this a poor use case. Reflection would be the way to go if we want to reduce this, where additional indirection would be the trade-off. I think it would be fine to use reflection in test code (i.e. here) but I think it would be best to avoid introducing it in production code, especially in this context (i.e. governance params).